### DISCUSSION TITLE
클로저에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#79] [클로저]
> 클로저에 대해 설명해주세요.
> lexical scope(렉시컬 스코프)가 없다면 클로저를 사용할 수 없나요?
> 클로저를 사용하는 이유는 무엇인가요?

**Comment by publdaze** [comment]
> - 클로저에 대해 설명해주세요.
- 클로저는 서로 중첩되어 있는 함수가 있을 때, 중첩 함수가 외부 함수의 식별자를 참조하고, 외부함수의 생명주기보다 중첩함수의 생명주기가 더 길 때에 해당하는 중첩함수를 뜻함
> - lexical scope(렉시컬 스코프)가 없다면 클로저를 사용할 수 없나요?
- 렉시컬 스코프는 함수가 호출되는 위치가 아닌 정의되는 위치를 기준으로 스코프를 따르는 것을 뜻함
- 만약 렉시컬 스코프가 없다면 클로저의 함수가 정의된 위치의 외부함수의 식별자에 참조할 수 없다는 것을 의미하기 때문에 일반적인 클로저의 정의로는 사용할 수 없음
> - 클로저를 사용하는 이유는 무엇인가요? 
- 클로저를 사용하는 이유는 일반적으로 함수 스코프 내의 식별자에 함수 외부에서 직접 접근하지 못하지만, 클로저를 사용하면 외부 함수의 생명주기가 끝났음에도 중첩함수를 통해 외부함수의 식별자에 접근이 가능
- 식별자의 직접 접근을 막되, 클로저에 반환된 중첩함수를 통한 캡슐화로 특정 동작을 가능하도록 처리할 수 있음
- 따라서 변수의 상태를 안전하게 관리할 수 있으며, 정보 은닉을 통해 실제 구현부는 숨기고, 메서드는 제공하여 사용을 할 수 있도록 처리하는 것도 가능하기 때문에 유용하게 사용할 수 있음

**Comment by pipisebastian** [comment]
> 클로저에 대해 설명해주세요.

- 내부함수가 상위 스코프인 외부함수의 식별자를 참조하고 있을때, 내부함수가 외부함수보다 더 오래 살아있어, 실행이 끝난 외부함수의 상위 스코프(렉시컬 환경)에 접근이 가능한겁니다.
- 내부함수가 외부함수의 렉시컬 환경을 참조하고 있기 때문에, 가비지 컬렉션의 대상이 되지 않고, 외부함수가 종료되더라도, 외부함수의 `렉시컬 환경` - `enviroment record` 가 살아있게 됩니다.

> lexical scope(렉시컬 스코프)가 없다면 클로저를 사용할 수 없나요?

- 클로저는 렉시컬 스코프가 없으면 제대로 작동할 수 없습니다. 클로저의 핵심 원리는 함수가 정의될 때의 스코프를 기억하고, 그 스코프에 접근할 수 있게 하는 것입니다. 이 원리는 렉시컬 스코프에 기반하고 있습니다

> 클로저를 사용하는 이유는 무엇인가요?

- 정보 은닉을 구현할 수 있습니다.
    - 정보 은닉이란, 외부에서 원치않는 값 변경을 막음으로써, 정보를 보호하고, 객체간의 의존성(결합도)를 낮춰줍니다.
